decoder/src/log/format.rs

```diff
@@ -1,48 +1,233 @@
+//! Formatting for `defmt` log frames.
+//!
+//! This module contains the parsers for the `defmt` custom log formatting strings.
+//!
+//! # Format strings
+//!
+//! A format string takes a set of [format specifiers] written
+//! in the way a log should be printed by the logger.
+//!
+//! ## Basics
+//!
+//! Format strings allow the customization of how the logger prints logs.
+//!
+//! The following log will be used as reference in the examples below:
+//!
+//! ```ignore
+//! defmt::error!("hello");
+//! ```
+//!
+//! The simplest format string is `"{s}"`. This prints the log and nothing else:
+//!
+//! ```text
+//! hello
+//! ```
+//!
+//! Arbitrary text can be added to the format string, which will be printed as specified with each log.
+//! For example, `"Log: {s}"`:
+//!
+//! ```text
+//! Log: hello
+//! ```
+//!
+//! Multiple specifiers can be included within a format string, in any order. For example `"[{L}] {s}"` prints:
+//!
+//! ```text
+//! [ERROR] hello
+//! ```
+//!
+//! ## Customizing log segments
+//!
+//! The way a format specifier is printed can be customized by providing additional, optional [format parameters].
+//!
+//! Format parameters are provided by adding the parameters after the format specifier, separated by colons (`:`),
+//! like this: `"{L:bold:5} {f:white:<10} {s}"`.
+//!
+//! ### Color
+//!
+//! A log segment can be specified to be colored by providing a color in the format parameters.
+//!
+//! There are three different options for coloring a log segment:
+//! - a string that can be parsed by the FromStr implementation of [colored::Color].
+//! - `severity` colors the log segment using the predefined color for the log level of log.
+//! - `werror` is similar to `severity`, but it only applies the color if the log level is WARN or ERROR.
```

Comment:
I am just wondering why we actually need `werror`. Isn't `severity` enough?

Reply:
I personally find that coloring warning and error logs is very useful, but coloring info, debug, and trace logs is way too noisy for me. Hence I added an option to color _just_ warnings and errors.

Reply:
Oki